File: 0110-balanced-binary-tree/main.go
package main

import ("fmt")
import ("math")


type TreeNode struct {
  Val int // 0
  Left *TreeNode // nil
  Right *TreeNode // nil
}

func depth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	left_depth := float64(depth(root.Left))
	right_depth := float64(depth(root.Right))
	abs_diff := math.Abs(left_depth - right_depth)
	if left_depth == -1 || right_depth == -1 || abs_diff > 1 {
		return -1
	}

	node_depth := math.Max(left_depth, right_depth) + 1
	return int(node_depth)
}

func isBalanced(root *TreeNode) bool {
	d := depth(root)
	return d >= 0
}

func main() {
	 // [1,2,2,3,3,null,null,4,4]

	e1 := TreeNode{
		Val: 1,
		Left: &TreeNode{
			Val: 2,
			Left: &TreeNode{
				Val: 3,
				Left: &TreeNode{Val: 4},
				Right: &TreeNode{Val: 4},
			},
			Right: &TreeNode{
				Val: 3,
			},
		},
		Right: &TreeNode{
			Val: 2,
		},
	}

	r  := isBalanced(&e1)

	fmt.Println(r)
}

File: 0110-balanced-binary-tree/main.js
// JavaScript program to check if a tree is height-balanced or not
// Using Bottom Up Recursion

class Node {
    constructor(d) {
        this.data = d;
        this.left = null;
        this.right = null;
    }
}

// https://www.geeksforgeeks.org/how-to-determine-if-a-binary-tree-is-balanced/
// https://www.geeksforgeeks.org/balanced-binary-tree/

var depth = function(root) {
  if (root === null) {
    return 0;
  }

  const left_depth = depth(root.left);
  const right_depth = depth(root.right);
  const invalid = -1;
  const is_invalid = (
    left_depth == invalid ||
    right_depth == invalid ||
    Math.abs(left_depth - right_depth) > 1
  );

  if (is_invalid)  {
    return invalid;
  }

  node_depth = Math.max(...[left_depth, right_depth]) + 1;
  return node_depth;
}

var isBalanced = function(root) {
  return depth(root) >= 0
};